src/shared/ai/langchain/langchain.service.ts

```diff
@@ -0,0 +1,103 @@
+import { Inject, Injectable, Logger } from '@nestjs/common';
+import { OpenAI } from 'langchain/llms/openai';
+import { ChatPromptTemplate } from 'langchain/prompts';
+import { ConversationChain } from 'langchain/chains';
+import { BufferMemory } from 'langchain/memory';
+import { OpenaiSecretsService } from '../openai/openai-secrets.service';
+import { Connection } from 'mongoose';
+import { getConnectionToken } from '@nestjs/mongoose';
+import { DatabaseConnectionService } from '../../database/database-connection.service';
+import { MongoDBChatMessageHistory } from '@langchain/community/stores/message/mongodb';
+import * as mongoose from 'mongoose';
+
+@Injectable()
+export class LangchainService {
+  private _model: OpenAI<any>;
+  private readonly _chatHistoryCollectionName = 'chat-history';
+
+  constructor(
+    @Inject(getConnectionToken()) private connection: Connection,
+    private readonly _openAiSecrets: OpenaiSecretsService,
+    private readonly _databaseConnectionService: DatabaseConnectionService,
+    private readonly _logger: Logger,
+  ) {}
+
+  /**
+   * Create a conversation chain that use MongoDBChatMessageHistory
+   * Based on example from here: https://js.langchain.com/docs/integrations/chat_memory/mongodb#usage
+   * @param modelName
+   */
+  async createMongoMemoryChatChain(modelName?: string) {
+    const model = await this._createModel(modelName);
+    const prompt = await this._createPrompt();
+    const memory = await this._createMemory();
+    return new ConversationChain({
+      llm: model,
+      prompt: prompt,
+      memory: memory,
+    });
+    this._logger.log(`Returning chain with model ${this._model.modelName}`);
+  }
```

Comment:
**Suggestion:** The `createMongoMemoryChatChain` method logs a message after returning the `ConversationChain` object. This log will never be executed. Consider moving the log statement before the return statement. [bug]
```suggestion
    this._logger.log(`Returning chain with model ${this._model.modelName}`);
    return new ConversationChain({
      llm: model,
      prompt: prompt,
      memory: memory,
    });
```